File: src/com/craftinginterpreters/lox/Environment.java
package com.craftinginterpreters.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Environment {
    private final Environment enclosing;
    private final Map<String, Object> valuesByName = new HashMap<>();
    private final Map<Integer, Object> valuesByIndex = new HashMap<>();

    Environment() {
        enclosing = null;
    }

    Environment(Environment enclosing) {
        this.enclosing = enclosing;
    }

    Object get(Token name) {
        if (valuesByName.containsKey(name.lexeme)) {
            return valuesByName.get(name.lexeme);
        }

        if (enclosing != null) {
            return enclosing.get(name);
        }

        throw new RuntimeError(name, "Undefined variable '" + name.lexeme + "'.");
    }

    Object getAt(int distance, Token name) {
        System.out.println("Environment::getAt - Looking up " + name.lexeme + " by name...");
        Environment anc = ancestor(distance);
        assert anc.valuesByName.containsKey(name.lexeme) : "Interpreter and Resolver out of sync - interpreter's environment should have variable" + name;
        return anc.valuesByName.get(name.lexeme);
    }

    Object getAt(int distance, int offset) {
        Environment anc = ancestor(distance);
        return anc.valuesByIndex.get(offset);
    }

    void define(String name, Object value) {
        valuesByName.put(name, value);
    }

    Environment ancestor(int distance) {
        Environment env = this;
        for (int i = 0; i < distance; i++) {
            env = env.enclosing;
        }
        return env;
    }

    void assign(Token name, Object value) {
        if (valuesByName.containsKey(name.lexeme)) {
            valuesByName.put(name.lexeme, value);
            return;
        }

        if (enclosing != null) {
            enclosing.assign(name, value);
            return;
        }

        throw new RuntimeError(name, "Undefined variable '" + name.lexeme + "'.");
    }

    void assignAt(int distance, Token name, int offset, Object value) {
        Environment env = ancestor(distance);
        env.valuesByName.put(name.lexeme, value);
        env.valuesByIndex.put(offset, value);
    }
}

File: rlox/src/environment.rs
use std::collections::HashMap;
use std::{cell::RefCell, rc::Rc};

use crate::error::RuntimeError;
use crate::object::LoxObject;
use crate::interpreter::Result;
use crate::scanner::Token;

#[derive(Clone)]
pub struct Environment(Rc<RefCell<EnvironmentData>>);

struct EnvironmentData {
    enclosing: Option<Environment>,
    values: HashMap<String, LoxObject>,
}

impl Environment {
    pub fn new() -> Self {
        Environment(Rc::new(RefCell::new(EnvironmentData {
            enclosing: None,
            values: HashMap::new(),
        })))
    }

    pub fn as_child_of(parent: Environment) -> Self {
        Environment(Rc::new(RefCell::new(EnvironmentData {
            enclosing: Some(parent),
            values: HashMap::new(),
        })))
    }

    pub fn define(&mut self, name: &str, value: &LoxObject) {
        self.0
            .borrow_mut()
            .values
            .insert(name.to_owned(), value.clone());
    }

    pub fn assign(&mut self, name: &Token, value: &LoxObject) -> Result<()> {
        if self.0.borrow().values.contains_key(&name.lexeme) {
            self.0
                .borrow_mut()
                .values
                .insert(name.lexeme.to_owned(), value.clone());
            Ok(())
        } else if let Some(enclosing) = &mut self.0.borrow_mut().enclosing {
            enclosing.assign(name, value)
        } else {
            Err(RuntimeError::new(
                name,
                &format!("Undefined variable \"{}\".", name.lexeme),
            ))
        }
    }

    pub fn get(&self, name: &Token) -> Result<LoxObject> {
        if let Some(v) = self.0.borrow().values.get(&name.lexeme) {
            Ok(v.clone())
        } else if let Some(parent) = &self.0.borrow().enclosing {
            parent.get(name)
        } else {
            Err(RuntimeError::new(
                name,
                &format!("Undefined variable \"{}\".", name.lexeme),
            ))
        }
    }

    pub fn get_at(&self, distance: usize, name: &str) -> Result<LoxObject> {
        if let Some(ancestor) = self.ancestor(distance) {
            if let Some(value) = ancestor.0.borrow().values.get(name) {
                Ok(value.clone())
            } else {
                Err(RuntimeError::with_message(&format!(
                    "Environment::get_at - Undefined variable \"{}\" at distance: {}.",
                    name, distance
                )))
            }
        } else {
            Err(RuntimeError::with_message(&format!(
                "No ancestor at distance \"{}\".",
                distance
            )))
        }
    }

    pub fn assign_at(&mut self, distance: usize, name: &Token, value: &LoxObject) -> Result<()> {
        if let Some(ancestor) = self.ancestor(distance) {
            ancestor
                .0
                .borrow_mut()
                .values
                .insert(name.lexeme.to_owned(), value.clone());
            Ok(())
        } else {
            Err(RuntimeError::new(
                name,
                &format!("No ancestor environment at distance \"{}\".", distance),
            ))
        }
    }

    fn ancestor(&self, distance: usize) -> Option<Self> {
        if distance == 0 {
            Some(self.clone())
        } else {
            let data = self.0.borrow();
            if let Some(enclosing) = &data.enclosing {
                return enclosing.ancestor(distance - 1);
            }
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::object::LoxObject;
    use crate::scanner::{Token, TokenType};

    #[test]
    fn define_works() {
        let mut env = Environment::new();
        let definitions = vec![
            (
                Token::new(TokenType::Identifier, String::from("a"), None, 1, 0),
                LoxObject::Number(10.0),
            ),
            (
                Token::new(TokenType::Identifier, String::from("b"), None, 1, 1),
                LoxObject::Str(String::from("Hello World")),
            ),
            (
                Token::new(TokenType::Identifier, String::from("c"), None, 1, 2),
                LoxObject::Boolean(false),
            ),
        ];

        for (name, value) in definitions {
            env.define(&name.lexeme, &value);
            assert_eq!(env.get(&name).unwrap(), value);
        }
    }

    #[test]
    fn assign_overwrites() {
        let mut env = Environment::new();
        let name = Token::new(TokenType::Identifier, String::from("a"), None, 1, 0);
        env.define(&name.lexeme, &LoxObject::Number(10.0));
        env.assign(&name, &LoxObject::Str(String::from("Hello World")))
            .unwrap();
        env.assign(&name, &LoxObject::Boolean(false)).unwrap();
        assert_eq!(env.get(&name).unwrap(), LoxObject::Boolean(false));
    }

    #[test]
    fn assign_cant_create_entry() {
        let mut env = Environment::new();
        let name = Token::new(TokenType::Identifier, String::from("a"), None, 1, 0);
        if let Ok(_) = env.assign(&name, &LoxObject::Nil) {
            panic!("Should not be able to assign to undefined variable.");
        }
    }

    #[test]
    fn define_overwrites() {
        let mut env = Environment::new();
        let name = Token::new(TokenType::Identifier, String::from("a"), None, 1, 0);
        env.define(&name.lexeme, &LoxObject::Number(10.0));
        env.define(&name.lexeme, &LoxObject::Str(String::from("Hello World")));
        env.define(&name.lexeme, &LoxObject::Boolean(false));

        assert_eq!(env.get(&name).unwrap(), LoxObject::Boolean(false));
    }
}
